### PR TITLE
dgram: make DATAGRAM an optional feature

### PR DESCRIPTION
**Pull request by LPardue**
This change makes the DATAGRAM extension an optional feature of
quiche. Conditional compilation of the library code and API is
controlled by two features "quic-dgram" and "h3-dgram". These
can be enabled when building quiche with e.g.
`cargo build --features "quic-dgram h3-dgram"`.

Since the features are optional, I've moved the dgram-client and
dgram-server example applications to their own crate "tools/dgram".
This crate enables the required features by default so it can just
be build with `cargo build`.

### CODE REVIEW

**Comment by LPardue** [comment]
@xanathar please take a look.

This is generally how the optional features would look, might requires some tweaks later when we attempt to get it mainlined.

File: src/dgram.rs [hunk 1]
```diff
@@ -23,20 +23,23 @@
 // LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 // NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 // SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-
+#[cfg(feature = "quic-dgram")]
```
Comment (xanathar) [inline src/dgram.rs:26]: Unless you have specific concerns, a single

`#![cfg(feature = "quic-dgram")]`

is likely more readable?

**Comment by xanathar** [comment]
Apart from the nitpick comment above about style, looks fine to me!

**Comment by LPardue** [comment]
weird I though I tried unssucessfully, but I tried it again and it worked, so I updated the PR and will merge

**Pull request merged**